core/src/main/java/org/apache/iceberg/SnapshotSummary.java

```diff
@@ -186,7 +187,13 @@ public Map<String, String> build() {
 
       metrics.addTo(builder);
       setIf(deletedDuplicateFiles > 0, builder, DELETED_DUPLICATE_FILES, deletedDuplicateFiles);
-      Set<String> changedPartitions = partitionMetrics.keySet();
+      // Exclude the key that corresponds to PartitionSpec.unpartitioned().
+      // We might want to keep it if and only if the table has had its partition spec updated over time and the
+      // write hit two specs (unpartitioned / VOID and an actual bound partition spec).
+      Set<String> changedPartitions = partitionMetrics.keySet()
+          .stream()
+          .filter(partKey -> partKey != null && !partKey.isEmpty())
+          .collect(Collectors.toSet());
```

Comment:
Why not just add a check on the line that adds partition summaries?

```java
  if (trustPartitionMetrics && isPartitioned && changedPartitions.size() <= maxChangedPartitionsForSummaries) {
    ...
  }
```

Reply:
I considered that, but we don't have access to `isPartitioned` here. What we know about the table's partitioning comes from the metrics and the partition keys.

I agree that's cleaner but requires more of a refactor. I'm going to update to simply not write a summary for the root partition (i.e. unpartitioned tables).

Reply:
And as mentioned elsewhere, we _do_ want to include some of the metrics about the root partition regardless (just not the summary).